File: backend/app/schemas/staff_leave.py
from __future__ import annotations

from decimal import Decimal
from datetime import date, datetime
from uuid import UUID
from pydantic import BaseModel, Field, ConfigDict


# ── Staff Leave Type Schemas ───────────────────────────────────────────────────

class StaffLeaveTypeBase(BaseModel):
    code: str = Field(..., max_length=30)
    name: str = Field(..., max_length=100)
    description: str | None = None
    max_days_per_year: Decimal = Field(default=Decimal("12.00"), ge=0)
    requires_attachment: bool = False
    is_paid: bool = True
    is_active: bool = True


class StaffLeaveTypeCreate(StaffLeaveTypeBase):
    pass


class StaffLeaveTypeUpdate(BaseModel):
    name: str | None = None
    description: str | None = None
    max_days_per_year: Decimal | None = None
    requires_attachment: bool | None = None
    is_paid: bool | None = None
    is_active: bool | None = None


class StaffLeaveTypeResponse(StaffLeaveTypeBase):
    id: UUID
    school_id: UUID
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


# ── Staff Leave Balance Schemas ────────────────────────────────────────────────

class StaffLeaveBalanceResponse(BaseModel):
    id: UUID
    school_id: UUID
    teacher_id: UUID
    academic_year_id: UUID
    leave_type_id: UUID
    leave_type_code: str
    leave_type_name: str
    allocated_days: Decimal
    used_days: Decimal
    pending_days: Decimal
    remaining_days: Decimal

    model_config = ConfigDict(from_attributes=True)


class StaffLeaveBalanceAllocate(BaseModel):
    teacher_id: UUID
    academic_year_id: UUID
    leave_type_id: UUID
    allocated_days: Decimal = Field(..., ge=0)


# ── Staff Leave Request Schemas ────────────────────────────────────────────────

class StaffLeaveRequestCreate(BaseModel):
    academic_year_id: UUID
    leave_type_id: UUID
    start_date: date
    end_date: date
    half_day_type: str = Field(default="FULL_DAY")  # FULL_DAY, FIRST_HALF, SECOND_HALF
    reason: str = Field(..., min_length=3)
    attachment_url: str | None = None


class StaffLeaveRequestUpdate(BaseModel):
    start_date: date | None = None
    end_date: date | None = None
    half_day_type: str | None = None
    reason: str | None = None
    attachment_url: str | None = None


class StaffLeaveApprovalAction(BaseModel):
    remarks: str | None = None


class StaffLeaveRejectAction(BaseModel):
    rejection_reason: str = Field(..., min_length=2)


class StaffLeaveApprovalHistoryResponse(BaseModel):
    id: UUID
    action_by_user_id: UUID
    action_by_name: str | None = None
    action: str
    from_status: str | None = None
    to_status: str
    remarks: str | None = None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class StaffLeaveRequestResponse(BaseModel):
    id: UUID
    school_id: UUID
    teacher_id: UUID
    teacher_name: str | None = None
    employee_id: str | None = None
    academic_year_id: UUID
    academic_year_name: str | None = None
    leave_type_id: UUID
    leave_type_code: str | None = None
    leave_type_name: str | None = None
    start_date: date
    end_date: date
    requested_days: Decimal
    half_day_type: str
    reason: str
    attachment_url: str | None = None
    status: str
    requested_by_user_id: UUID | None = None
    requested_by_name: str | None = None
    approved_by_user_id: UUID | None = None
    approved_by_name: str | None = None
    rejected_by_user_id: UUID | None = None
    rejected_by_name: str | None = None
    cancelled_by_user_id: UUID | None = None
    approval_remarks: str | None = None
    rejection_reason: str | None = None
    approved_at: datetime | None = None
    rejected_at: datetime | None = None
    cancelled_at: datetime | None = None
    created_at: datetime
    updated_at: datetime
    approval_history: list[StaffLeaveApprovalHistoryResponse] = []

    model_config = ConfigDict(from_attributes=True)


class StaffLeaveSummaryReport(BaseModel):
    academic_year_id: UUID
    total_requests: int
    pending_requests: int
    approved_today: int
    currently_on_leave: int
    by_leave_type: dict[str, int]
    by_status: dict[str, int]

File: frontend/src/api/staffLeave.ts
import { apiClient } from '@/services/api/client';

export interface StaffLeaveType {
  id: string;
  school_id: string;
  code: string;
  name: string;
  description?: string;
  max_days_per_year: number;
  requires_attachment: boolean;
  is_paid: boolean;
  is_active: boolean;
}

export interface StaffLeaveBalance {
  id: string;
  school_id: string;
  teacher_id: string;
  academic_year_id: string;
  leave_type_id: string;
  leave_type_code: string;
  leave_type_name: string;
  allocated_days: number;
  used_days: number;
  pending_days: number;
  remaining_days: number;
}

export interface StaffLeaveApprovalHistory {
  id: string;
  action_by_user_id: string;
  action_by_name?: string;
  action: string;
  from_status?: string;
  to_status: string;
  remarks?: string;
  created_at: string;
}

export interface StaffLeaveRequest {
  id: string;
  school_id: string;
  teacher_id: string;
  teacher_name?: string;
  employee_id?: string;
  academic_year_id: string;
  academic_year_name?: string;
  leave_type_id: string;
  leave_type_code?: string;
  leave_type_name?: string;
  start_date: string;
  end_date: string;
  requested_days: number;
  half_day_type: string;
  reason: string;
  attachment_url?: string;
  status: string;
  requested_by_user_id?: string;
  requested_by_name?: string;
  approved_by_user_id?: string;
  approved_by_name?: string;
  rejected_by_user_id?: string;
  rejected_by_name?: string;
  cancelled_by_user_id?: string;
  approval_remarks?: string;
  rejection_reason?: string;
  approved_at?: string;
  rejected_at?: string;
  cancelled_at?: string;
  created_at: string;
  updated_at: string;
  approval_history: StaffLeaveApprovalHistory[];
}

export interface StaffLeaveSummaryReport {
  academic_year_id: string;
  total_requests: number;
  pending_requests: number;
  approved_today: number;
  currently_on_leave: number;
  by_leave_type: Record<string, number>;
  by_status: Record<string, number>;
}

export const staffLeaveApi = {
  getLeaveTypes: async (): Promise<StaffLeaveType[]> => {
    const res = await apiClient.get('/api/v1/staff-leave/types');
    return res.data.data;
  },

  createLeaveType: async (data: Partial<StaffLeaveType>): Promise<StaffLeaveType> => {
    const res = await apiClient.post('/api/v1/staff-leave/types', data);
    return res.data.data;
  },

  updateLeaveType: async (typeId: string, data: Partial<StaffLeaveType>): Promise<StaffLeaveType> => {
    const res = await apiClient.put(`/api/v1/staff-leave/types/${typeId}`, data);
    return res.data.data;
  },

  getMyBalances: async (academicYearId: string): Promise<StaffLeaveBalance[]> => {
    const res = await apiClient.get('/api/v1/staff-leave/balance', { params: { academic_year_id: academicYearId } });
    return res.data.data;
  },

  getTeacherBalances: async (teacherId: string, academicYearId: string): Promise<StaffLeaveBalance[]> => {
    const res = await apiClient.get(`/api/v1/staff-leave/balance/${teacherId}`, { params: { academic_year_id: academicYearId } });
    return res.data.data;
  },

  allocateBalance: async (data: { teacher_id: string; academic_year_id: string; leave_type_id: string; allocated_days: number }): Promise<StaffLeaveBalance> => {
    const res = await apiClient.post('/api/v1/staff-leave/balances/allocate', data);
    return res.data.data;
  },

  getLeaveRequests: async (params?: {
    teacher_id?: string;
    status?: string;
    leave_type_id?: string;
    start_date?: string;
    end_date?: string;
    academic_year_id?: string;
  }): Promise<StaffLeaveRequest[]> => {
    const res = await apiClient.get('/api/v1/staff-leave', { params });
    return res.data.data;
  },

  getLeaveRequestById: async (id: string): Promise<StaffLeaveRequest> => {
    const res = await apiClient.get(`/api/v1/staff-leave/${id}`);
    return res.data.data;
  },

  createLeaveRequest: async (data: {
    academic_year_id: string;
    leave_type_id: string;
    start_date: string;
    end_date: string;
    half_day_type?: string;
    reason: string;
    attachment_url?: string;
  }): Promise<StaffLeaveRequest> => {
    const res = await apiClient.post('/api/v1/staff-leave', data);
    return res.data.data;
  },

  approveLeaveRequest: async (id: string, remarks?: string): Promise<StaffLeaveRequest> => {
    const res = await apiClient.post(`/api/v1/staff-leave/${id}/approve`, { remarks });
    return res.data.data;
  },

  rejectLeaveRequest: async (id: string, rejection_reason: string): Promise<StaffLeaveRequest> => {
    const res = await apiClient.post(`/api/v1/staff-leave/${id}/reject`, { rejection_reason });
    return res.data.data;
  },

  cancelLeaveRequest: async (id: string, remarks?: string): Promise<StaffLeaveRequest> => {
    const res = await apiClient.post(`/api/v1/staff-leave/${id}/cancel`, { remarks });
    return res.data.data;
  },

  getSummaryReport: async (academicYearId: string): Promise<StaffLeaveSummaryReport> => {
    const res = await apiClient.get('/api/v1/staff-leave/reports/summary', { params: { academic_year_id: academicYearId } });
    return res.data.data;
  },
};
